### PR TITLE
`Adaptive learning`: Improve prerequisite import

### PR DESCRIPTION
**Pull request by rstief**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
We want to refactor the prerequisite management with the goal of linking exercise units to prerequisites as `pre-tests` for a course. To support this, prerequisites need to be changed to be actual entities (before they were only references).

### Description
This PR adds Prerequisites as entities to the server:
- Adds a new entities  `Prerequisite` and `CourseCompetency` (which is the new superclass of `Competency` and `Prerequisite`)
- Renames the table `competency` to `course_competency` as it now contains all `CourseCompetencies`.
- Migrates all prerequisites from the old table to the `course_competency` table

It also improves the prerequisite import:
- Prerequisites are no longer imported one-by-one
- Multiple competencies can be selected at once to be imported as prerequisites.

Other changes:
- Changes the `Competency` class in the client to be an interface -> to support future use cases where we want to use the new `CourseCompetency` interface for both competencies and prerequisites.

### Steps for Testing
Needed:
- 1 Instructor
- Another course with competencies (you can use Raphael Stief Test Course)

- Test the new prerequisite import (select, de-select competencies, search, filter, import, ...)
- Test the competency import (import from another course)
- Test all other occurrences of prerequisites:
  - Competency Management Page
  - View Prerequisites when Self-enrolling

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| competency-management.component.ts | 94.21% | ✅  |
| competency.service.ts | 90.54% | ✅  |
| create-competency.component.ts | 83.33% | ✅  |
| competency-search.component.ts | 100% | ✅ |
| import-competencies.component.ts | 90.9% | ✅  |
| import-course-competencies.component.ts | 94.36% | ✅  |
| import-prerequisites.component.ts | 91.66% | ✅  |
| prerequisite.service.ts | 100% | ✅ |
| course-management.route.ts | not found (modified) |  |
| competency.model.ts | 87.93% | ✅  |
| standardized-competency.model.ts | 100% | ✅ |
| course.model.ts | 100% | ✅ |
| lecture-wizard-competencies.component.ts | 97.24% | ✅  |
| course-competencies.component.ts | 95.31% | ✅  |
| course-prerequisites-modal.component.ts | 94.73% | ✅ |
| pageable-table.ts | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| Course.java | 92% | ✅  |
| BaseCompetency.java | 100% | ✅ |
| Competency.java | 75% | ✅ |
| CourseCompetency.java | 100% | ✅ |
| Prerequisite.java | 25% | ✅ (only constructor) |
| CompetencyRepository.java | 63% | ✅ (no methods added) |
| CourseCompetencyRepository.java | 84% | ✅ |
| PrerequisiteRepository.java | 100% | ✅ |
| CourseService.java | 93% | ✅  |
| CompetencyService.java | 92% | ✅  |
| PrerequisiteService.java | 100% | ✅ |
| CompetencyResource.java | 94% | ✅  |
| PrerequisiteResource.java | 100% | ✅ |
| PrerequisiteResponseDTO.java | 100% | ✅ |

### Screenshots
image

## Summary by CodeRabbit

- **New Features**
  - Introduced functionality for importing and managing course competencies and prerequisites in the user interface.
  - Added `Prerequisite` and `LinkedCourseCompetencyDTO` classes to enhance competency handling.

- **Bug Fixes**
  - Improved conditional rendering and null handling in competency-related HTML templates.

- **Documentation**
  - Updated i18n files to refine button labels, success messages, and text related to competency import and management.

- **Refactor**
  - Restructured `Competency` class into interfaces for better maintainability and updated related services and repositories.

- **Tests**
  - Enhanced test coverage for competency creation, prerequisites handling, and importing functionalities.

- **Chores**
  - Removed outdated methods and adjusted security annotations for streamlined access control in competency management.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/competency/PrerequisiteResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/competency/PrerequisiteResource.java:64]:
```suggestion
    @EnforceAtLeastStudentInCourse
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Code looks really good overall. Found one small thing:

File: src/main/java/de/tum/in/www1/artemis/web/rest/competency/PrerequisiteResource.java
Comment (rstief) [reply src/main/java/de/tum/in/www1/artemis/web/rest/competency/PrerequisiteResource.java:64]: This is actually intended. We want to allow students to view the prerequisites of courses they can enroll into.
 If you look at the code below:
```
if (!course.isEnrollmentEnabled()) {
    authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, null);
}
```
We only want to check if the student is in this course **if the self enrollment is not enabled**.

File: src/main/java/de/tum/in/www1/artemis/web/rest/competency/PrerequisiteResource.java
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/web/rest/competency/PrerequisiteResource.java:62]:
```suggestion
    /**
     * GET /courses/:courseId/competencies/prerequisites : Gets all prerequisite competencies for a course.
     * This endpoint allows all students to view prerequisites of a course if self-enrollment is activated (and thus only uses @EnforceAtLeastStudent)
     *
     * @param courseId the id of the course for which the prerequisites should be fetched for
     * @return the ResponseEntity with status 200 (OK) and with body the found prerequisites
     */
```

**Review by sarpsahinalp — COMMENTED** [review]
Tested locally. When there is no element in the search table on the Prerequisite Import Page and the Competency Import Page, the following message is displayed: "Translation not found."
image

File: src/main/webapp/app/course/competencies/import-competencies/import-course-competencies.component.html
Comment (rstief) [inline src/main/webapp/app/course/competencies/import-competencies/import-course-competencies.component.html:12]:
```suggestion
        <span jhiTranslate="'artemisApp.competency.import.searchTableEmpty'"></span>
```

**Review by JohannesStoehr — APPROVED** [review]
Code

File: src/main/java/de/tum/in/www1/artemis/web/rest/competency/PrerequisiteResource.java
Comment (Jan-Thurner) [reply src/main/java/de/tum/in/www1/artemis/web/rest/competency/PrerequisiteResource.java:65]: I am not sure how this works: If one was not enrolled in a single course, would one be a student? What I want to say is: If one was enrolling in a course for the first time, would `@EnforceAtLeastStudent` permit them to see the prerequisites? 

If one has the role `user` until they first enrolled in a course, this could be an edge case to consider. If the role `student` is given to anyone who signed in to Artemis, this should not be an issue.

Comment (rstief) [reply src/main/java/de/tum/in/www1/artemis/web/rest/competency/PrerequisiteResource.java:65]: Good point @Jan-Thurner . I looked up the annotation, it checks against the role "user" so this should be no problem (see `EnforceAtLeastStudent.java`)

image

File: src/main/webapp/i18n/de/competency.json
Comment (Jan-Thurner) [inline src/main/webapp/i18n/de/competency.json:196]:
```suggestion
                "importButton": "Voraussetzungen importieren",
```

File: src/main/webapp/i18n/de/competency.json
Comment (Jan-Thurner) [inline src/main/webapp/i18n/de/competency.json:197]:
```suggestion
                "importFromCoursesButton": "Aus anderen Kursen importieren",
```

**Review by Jan-Thurner — CHANGES_REQUESTED** [review]
Except for two missing translations, the Code lgtm. I also left a comment on @JohannesStoehr 's remark.

File: src/main/webapp/app/course/competencies/create-competency/create-competency.component.ts
Comment (JohannesWt) [inline src/main/webapp/app/course/competencies/create-competency/create-competency.component.ts:35]: This should not be needed as in competencyToCreate is already initialised

File: src/main/webapp/app/course/competencies/prerequisite.service.ts
Comment (JohannesWt) [inline src/main/webapp/app/course/competencies/prerequisite.service.ts:51]: this could be made static (optional: could be moved into Prerequisite by changing it to a class)

**Review by JohannesWt — CHANGES_REQUESTED** [review]
code looks good - just two small comments

File: src/main/webapp/app/course/competencies/create-competency/create-competency.component.ts
Comment (rstief) [inline src/main/webapp/app/course/competencies/create-competency/create-competency.component.ts:35]:
```suggestion
```

File: src/main/webapp/app/course/competencies/prerequisite.service.ts
Comment (rstief) [inline src/main/webapp/app/course/competencies/prerequisite.service.ts:51]:
```suggestion
    private static convertResponseDTOToPrerequisite(prerequisiteDTO: PrerequisiteResponseDTO): Prerequisite {
```

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by JohannesWt — DISMISSED** [review]
Code 👍

**Review by Jan-Thurner — DISMISSED** [review]
Code now lgtm

**Review by sarpsahinalp — DISMISSED** [review]
It seems that a server test is failing, and the code coverage report gives the following error for the client side:
Jest: "global" coverage threshold for branches (73.23%) not met: 73.21%.

**Comment by rstief** [comment]
> It seems that a server test is failing, and the code coverage report gives the following error for the client side: Jest: "global" coverage threshold for branches (73.23%) not met: 73.21%.

The server test is unrelated. I added additional coverage to the client, I hope it is now sufficient.

File: src/test/javascript/spec/component/competencies/prerequisite.service.spec.ts
Comment (rstief) [reply src/test/javascript/spec/component/competencies/prerequisite.service.spec.ts:101]: convertResponseDTOToPrerequisite is a private method. Do you still think it is a good idea?

**Review by coolchock — APPROVED** [review]
tested during testing session, works as expected 👍

**Review by EneaGore — APPROVED** [review]
Tested in testing session TS2. Looks good.

**Review by BBesrour — DISMISSED** [review]
Tested in a testing session and manually on TS2. Works as described

**Review by muradium — DISMISSED** [review]
Tested on TS2. Works as expected

**Review by EneaGore — APPROVED** [review]
Works as expected. Tested on TS2

**Review by Jan-Thurner — APPROVED** [review]
New Changes LGTM

File: src/main/webapp/app/course/competencies/competency-management/competency-management.component.ts
Comment (JohannesWt) [inline src/main/webapp/app/course/competencies/competency-management/competency-management.component.ts:69]: You could declare all services as `readonly`

File: src/main/webapp/app/course/competencies/prerequisite.service.ts
Comment (JohannesWt) [inline src/main/webapp/app/course/competencies/prerequisite.service.ts:20]: How can there be no response body? Is there no exception from the server then?

**Review by JohannesWt — CHANGES_REQUESTED** [review]
Code looks code - just one two comments/questions

File: src/main/webapp/app/course/competencies/prerequisite.service.ts
Comment (rstief) [reply src/main/webapp/app/course/competencies/prerequisite.service.ts:20]: I just looked it up and you're right, the server always returns an array.

**Review by muradium — DISMISSED** [review]
Changes LGTM

**Review by JohannesWt — DISMISSED** [review]
Thanks for the changes - Code Looks good

**Review by coolchock — DISMISSED** [review]
re-approve

**Review by muradium — DISMISSED** [review]
Reapprove

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove code

**Review by Jan-Thurner — APPROVED** [review]
reapprove code

**Review by BBesrour — APPROVED** [review]
re-approve

File: src/main/java/de/tum/in/www1/artemis/domain/competency/Prerequisite.java
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/domain/competency/Prerequisite.java:10]: I'd like to see some documentation here

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/competency/PrerequisiteResponseDTO.java
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/competency/PrerequisiteResponseDTO.java:36]: Why is this a private DTO? Why not a separate record that's reusable by others?

File: src/test/java/de/tum/in/www1/artemis/competency/PrerequisiteUtilService.java
Comment (MaximilianAnzinger) [inline src/test/java/de/tum/in/www1/artemis/competency/PrerequisiteUtilService.java:14]: Java Doc

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/competency/PrerequisiteResponseDTO.java
Comment (rstief) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/competency/PrerequisiteResponseDTO.java:36]: Fair

File: src/main/webapp/app/course/competencies/import-competencies/import-course-competencies.component.html
Comment (rstief) [inline src/main/webapp/app/course/competencies/import-competencies/import-course-competencies.component.html:12]:
```suggestion
        <span jhiTranslate="artemisApp.competency.import.searchTableEmpty"></span>
```

Comment (rstief) [reply src/main/webapp/app/course/competencies/import-competencies/import-course-competencies.component.html]: fixed a translation that was broken due to additional quotation marks :)

**Review by BBesrour — DISMISSED** [review]
Changes lgtm!
Re-approve

File: src/test/java/de/tum/in/www1/artemis/competency/PrerequisiteUtilService.java
Comment (rstief) [reply src/test/java/de/tum/in/www1/artemis/competency/PrerequisiteUtilService.java:63]: this is for tests only.

**Review by JohannesStoehr — COMMENTED** [review]
Code reapprove

**Review by JohannesWt — APPROVED** [review]
Reapprove - Code

**Review by MaximilianAnzinger — DISMISSED** [review]
Maintainer approved

**Review by krusche — APPROVED** [review]
I tested the migration with production data and it works fine 👍

**Review by JohannesStoehr — APPROVED** [review]
Reapprove Code after merge

**Review by Jan-Thurner — APPROVED** [review]
tested again on TS1, works as expected.

**Review by EneaGore — APPROVED** [review]
Manual test on TS1. Lgtm

**Review by BBesrour — APPROVED** [review]
new changes lgtm! (re-approve)

**Pull request merged**
